morl_baselines/multi_policy/pareto_q_learning/pql.py

```diff
@@ -1,312 +1,244 @@
+        """Learn the Pareto front.
 
         Args:
-            iterations (int, optional): The number of iterations to execute. (Default value = 3000)
-            max_timesteps (int, optional): The maximum timesteps for each iteration. (Default value = 100)
-            log (bool, optional): Whether to log the results or not. (Default value = True)
-            log_every (int, optional): Log the performance metric every number of iterations. (Default value = 1)
-            project_name (str, optional): The name for the wandb project. (Default value = "PQL")
-            experiment_name (str, optional): The name for the wandb experiment. (Default value = "PQL")
+            num_episodes (int, optional): The number of episodes to train for.
+            max_timesteps (int, optional): The maximum number of time steps in a single episode. (Default value = 1000)
+            log_every (int, optional): Log the results every number of episodes. (Default value = 100)
+            action_eval (str, optional): The action evaluation function name. (Default value = 'hypervolume')
 
         Returns:
-            /
+            Set: The final Pareto front.
         """
+        if action_eval == 'hypervolume':
+            score_func = self.score_hypervolume
+        elif action_eval =='pareto_cardinality':
+            score_func = self.score_pareto_cardinality
+        else:
+            raise Exception('No other method implemented yet')
 
-        if log:
-            self.setup_wandb(project_name, experiment_name)
+        for episode in range(num_episodes):
+            if episode % log_every == 0:
+                print(f'Training episode {episode + 1}')
 
-        for i in range(iterations):
-            obs, _ = self.env.reset()
-            state = self.flatten_observation(obs)
+            state, _ = self.env.reset()
+            state = int(np.ravel_multi_index(state, self.env_shape))
             terminated = False
             truncated = False
             timestep = 0
 
             while not (terminated or truncated) and timestep < max_timesteps:
-                action = self.select_action(state)
-                next_state, r, terminated, truncated, prob = self.env.step(action)
-                next_state = self.flatten_observation(next_state)
-                self.update(state, action, next_state, r)
+                action = self.select_action(state, score_func)
+                next_state, reward, terminated, truncated, _ = self.env.step(action)
+                next_state = int(np.ravel_multi_index(next_state, self.env_shape))
+
+                self.counts[state, action] += 1
+                self.non_dominated[state][action] = self.calc_non_dominated(next_state)
+                self.avg_reward[state, action] += (reward - self.avg_reward[state, action]) / self.counts[state, action]
+                self.epsilon = max(self.final_epsilon, self.epsilon * self.epsilon_decay)
```

Comment:
Epsilon is decayed exponentially in the original paper.
![image](https://user-images.githubusercontent.com/11799929/204253969-55b21ee7-f864-4a75-ab97-503896aa052e.png)

Reply:
I do that every episode, so it is exponential. For example, in episode 5, epsilon is already: ```epsilon = decay * decay * decay * decay * epsilon```.

Reply:
I did notice that I updated epsilons inside single episodes, so I took that out and now do it only after having concluded a full episode.